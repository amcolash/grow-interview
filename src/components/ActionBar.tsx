import React from 'react';
import { style } from 'typestyle';

const barStyle = style({
  display: 'flex',
  width: '100%',
  height: '96px',
  padding: '16px',
  justifyContent: 'center',
  alignItems: 'center',
  gap: '16px',

  borderRadius: '100px',
  background: 'var(--neutral-000, #FFF)',
  boxShadow: '0px 2px 0px 1px rgba(5, 9, 13, 0.06)',
});

export function ActionBar() {
  return <div className={barStyle}></div>;
}
