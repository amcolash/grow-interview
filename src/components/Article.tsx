import React from 'react';
import { style } from 'typestyle';

interface ArticleProps {
  index: number;
  title: string;
  views: number;
}

const articleStyle = style({
  display: 'flex',
  padding: '24px',
  alignItems: 'center',
  gap: '20px',
  alignSelf: 'stretch',
  fontSize: '16px',
  lineHeight: '24px',

  borderRadius: '12px',
  border: '1px solid var(--neutral-300, #E7EAEB)',
  background: 'var(--white, #FFF)',

  $nest: {
    '& .index, & .views': {
      color: 'var(--neutral-500, #A7AAAB)',
    },
    '.title': {
      flex: 1,
    },
    '.views': {
      fontSize: '14px',
      fontFamily: 'Poppins',
      lineHeight: '16px',
    },
  },
});

export function Article(props: ArticleProps) {
  return (
    <article className={articleStyle}>
      <div className="index">{props.index}</div>
      <div className="title">{props.title}</div>
      <div className="views">{props.views.toLocaleString('en-US')} views</div>
    </article>
  );
}
