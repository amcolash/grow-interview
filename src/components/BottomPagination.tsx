import React from 'react';
import ReactPaginate from 'react-paginate';
import { style } from 'typestyle';

import { ReactComponent as ChevronLeft } from '../icons/chevron-left.svg';
import { ReactComponent as ChevronRight } from '../icons/chevron-right.svg';

interface BottomPaginationProps {
  totalPages: number;
  currentPage: number;
  onPageChange: (selectedItem: { selected: number }) => void;
}

const pageListStyle = style({
  display: 'flex',
  justifyContent: 'center',
  flexWrap: 'wrap',
  gap: '8px',
  listStyle: 'none',
  margin: 0,
  padding: 0,

  $nest: {
    '& a': {
      display: 'flex',
      justifyContent: 'center',
      alignItems: 'center',
      width: '40px',
      height: '40px',
      padding: '10px',
      color: 'var(--neutral-900, #05090D)',
      fontFamily: 'Poppins',
    },
    '& li': {
      border: '1px solid var(--neutral-400, #D4D8D9)',
      background: 'var(--neutral-000, #FFF)',
      borderRadius: '100px',
      userSelect: 'none',

      $nest: {
        '&.disabled': {
          background: 'var(--neutral-400, #D4D8D9)',
        },
        '&.disabled > a': {
          color: 'var(--neutral-600, #737680)',
        },
        '&:not(.disabled,.selected):hover': {
          background: 'var(--neutral-300, #e7eaeb)',
        },
        '&.selected': {
          background: 'var(--avocado-300, #E0E9CB)',
        },
        '&.selected > a': {
          color: 'var(--brand-green-500, #025B4B)',
        },
      },
    },
  },
});

export function BottomPagination(props: BottomPaginationProps) {
  return (
    <ReactPaginate
      className={pageListStyle}
      breakLabel="..."
      nextLabel={<ChevronRight />}
      onPageChange={props.onPageChange}
      pageCount={props.totalPages}
      previousLabel={<ChevronLeft />}
      renderOnZeroPageCount={null}
    />
  );
}
