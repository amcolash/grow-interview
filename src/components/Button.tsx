import React, { ButtonHTMLAttributes } from 'react';
import { media, style } from 'typestyle';

import { mobileWidth } from '../consts';

const buttonStyle = style(
  {
    display: 'flex',
    width: '160px',
    padding: '12px 24px',
    justifyContent: 'center',
    alignItems: 'center',
    gap: '8px',
    flexShrink: '0',
    alignSelf: 'stretch',
    borderRadius: '100px',
    background: 'var(--brand-green-500, #025b4b)',
    border: 'none',
    color: 'var(--neutral-000, #fff)',
    textAlign: 'center',
    fontSize: '16px',
    fontFamily: 'Poppins',
    fontWeight: '500',
    lineHeight: '24px',
    letterSpacing: '-0.32px',
  },
  media({ maxWidth: mobileWidth }, { width: '100%' })
);

export function Button(props: ButtonHTMLAttributes<HTMLButtonElement> & { 'data-testid'?: string }) {
  return <button {...props} className={buttonStyle} data-testid={props['data-testid'] || 'button'} />;
}
