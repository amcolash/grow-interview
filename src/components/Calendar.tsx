import React from 'react';
import { CaptionProps, DayPicker, useNavigation } from 'react-day-picker';
import 'react-day-picker/dist/style.css';
import { cssRule, style } from 'typestyle';

import { ReactComponent as ChevronLeft } from '../icons/chevron-left.svg';
import { ReactComponent as ChevronRight } from '../icons/chevron-right.svg';

interface CalendarProps {
  date: Date;
  onChange: (date: Date) => void;
}

cssRule('.rdp', {
  fontFamily: 'Poppins',
  margin: 0,
});

export function Calendar(props: CalendarProps) {
  return (
    <DayPicker
      mode="single"
      selected={props.date}
      onDayClick={(day) => props.onChange(day)}
      disabled={{ from: new Date(), to: new Date(9999, 12, 31) }}
      formatters={{
        formatWeekdayName: (weekday) =>
          weekday.toLocaleDateString(undefined, {
            weekday: 'short',
          }),
      }}
      styles={{
        head_cell: {
          fontWeight: 'normal',
          color: 'var(--neutral-500, #A7AAAB)',
          fontSize: '12px',
          lineHeight: '18px',
        },
        cell: {
          fontSize: '14px',
          lineHeight: '22px',
        },
      }}
      components={{
        Caption: CustomCaption,
      }}
      modifiersStyles={{
        selected: {
          color: 'var(--brand-green-500, #025B4B)',
          background: 'var(--ivy-300, #B3CEC9)',
          fontWeight: '500',
        },
      }}
    />
  );
}

const captionStyle = style({
  display: 'flex',
  justifyContent: 'space-between',
  marginBottom: '32px',

  $nest: {
    '& button': {
      border: 'none',
      background: 'none',
      display: 'flex',
      padding: '0',
      alignItems: 'center',
      width: '24px',
      height: '24px',
      justifyContent: 'center',
      borderRadius: '100%',
    },
    '& button:hover': {
      background: 'var(--neutral-300, #e7eaeb)',
    },
    '& label': {
      fontSize: '16px',
      lineHeight: '24px',
      fontWeight: '500',
    },
  },
});

function CustomCaption(props: CaptionProps) {
  const { previousMonth, nextMonth, goToMonth, displayMonths } = useNavigation();
  const caption = props.displayMonth.toLocaleDateString(undefined, {
    month: 'long',
    year: 'numeric',
  });

  return (
    <div className={captionStyle}>
      <button
        onClick={() => {
          if (!previousMonth) return;
          goToMonth(previousMonth);
        }}
      >
        <ChevronLeft />
      </button>
      <label>{caption}</label>
      <button
        onClick={() => {
          if (!nextMonth) return;
          goToMonth(nextMonth);
        }}
        disabled={nextMonth ? nextMonth.getTime() > Date.now() : false}
      >
        <ChevronRight />
      </button>
    </div>
  );
}
