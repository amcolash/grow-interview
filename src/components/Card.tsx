import React from 'react';
import { style } from 'typestyle';

interface CardProps {
  children: React.ReactNode;
}

const cardListStyle = style({
  display: 'flex',
  flexDirection: 'column',
  gap: '20px',
  padding: '32px',

  borderRadius: '16px',
  background: '#FFF',
  boxShadow: '0px 2px 0px 1px rgba(5, 9, 13, 0.06)',
});

export function Card(props: CardProps) {
  return <div className={cardListStyle}>{props.children}</div>;
}
