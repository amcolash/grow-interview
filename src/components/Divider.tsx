import React from 'react';
import { media, style } from 'typestyle';

import { mobileWidth } from '../consts';

const dividerStyle = style(
  {
    borderLeft: '1px solid var(--neutral-300, #E7EAEB)',
    height: '64px',
    margin: '0 16px',
  },
  media({ maxWidth: mobileWidth }, { display: 'none' })
);

export function Divider() {
  return <div className={dividerStyle}></div>;
}
