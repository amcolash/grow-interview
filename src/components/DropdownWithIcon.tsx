import { Menu } from '@szhsin/react-menu';
import '@szhsin/react-menu/dist/transitions/slide.css';
import React, { FormEvent, MouseEventHandler } from 'react';
import { classes, media, style } from 'typestyle';

import { ReactComponent as ChevronDown } from '../icons/chevron-down.svg';
import { ReactComponent as ChevronUp } from '../icons/chevron-up.svg';

import { maxWidth } from '../consts';

interface DropdownWithIconProps {
  icon: React.ReactNode;
  label: string;
  value: React.ReactNode;
  children: React.ReactNode;
  onChange?: (value: any) => void;
  menuClassName?: string;
}

const wrapperStyle = style(
  {
    display: 'flex',
    padding: '12px',
    alignItems: 'center',
    gap: '24px',
    flex: '1 0 0',
    fontFamily: 'Poppins',
    borderRadius: '100px',
    userSelect: 'none',
    border: 'none',
    background: 'none',
    minWidth: '220px',

    $nest: {
      '&:hover, &:active': {
        background: 'var(--neutral-100, #F5F7F7)',
      },
      '.text': {
        display: 'flex',
        flexDirection: 'column',
        alignItems: 'flex-start',
        gap: '2px',
      },
      '.label': {
        color: 'var(--neutral-500, #A7AAAB)',
        fontSize: '12px',
        fontWeight: '500',
        lineHeight: '16px',
        letterSpacing: '0.6px',
        textTransform: 'uppercase',

        $nest: {
          label: {
            marginRight: '6px',
          },
        },
      },
      '.value': {
        color: 'var(--black, #000)',
        fontSize: '16px',
        lineHeight: '24px',
        letterSpacing: '-0.32px',
      },
    },
  },
  media({ maxWidth }, { padding: '2px' })
);

const menuStyle = style({
  display: 'flex',
  padding: '32px 16px',
  flexDirection: 'column',
  alignItems: 'flex-start',
  gap: '24px',
  borderRadius: '24px',
  background: 'var(--neutral-000, #FFF)',
  boxShadow: '0px 4px 24px 0px rgba(0, 0, 0, 0.12)',
  userSelect: 'none',
});

export function DropdownWithIcon(props: DropdownWithIconProps) {
  return (
    <Menu
      menuButton={({ open }) => <MenuButton {...props} open={open} />}
      onItemClick={(event) => {
        if (props.onChange) props.onChange(event.value);
      }}
      menuClassName={classes(menuStyle, props.menuClassName)}
      gap={10}
      transition
    >
      {props.children}
    </Menu>
  );
}

/** Internal Button Component */
interface MenuButtonProps {
  icon: React.ReactNode;
  label: string;
  value: React.ReactNode;
  open: boolean;
  onClick?: MouseEventHandler<HTMLButtonElement> | undefined;
}

const MenuButton = React.forwardRef<HTMLButtonElement, MenuButtonProps>(function Internal(props, ref) {
  return (
    <button className={wrapperStyle} ref={ref} onClick={props.onClick}>
      {props.icon}

      <div className="text">
        <div className="label">
          <label>{props.label}</label>
          {props.open ? <ChevronUp /> : <ChevronDown />}
        </div>

        {props.value}
      </div>
    </button>
  );
});
