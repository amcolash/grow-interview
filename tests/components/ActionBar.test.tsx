import { render } from '@testing-library/react';
import React from 'react';

import { ActionBar } from '../../src/components/ActionBar';

describe('ActionBar', () => {
  it('renders', () => {
    const { asFragment } = render(
      <ActionBar
        searchDate={new Date('6/26/2023')}
        language="gb"
        pageSize={100}
        setLanguage={() => {}}
        setPageSize={() => {}}
        setSearchDate={() => {}}
        setUrl={() => {}}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
