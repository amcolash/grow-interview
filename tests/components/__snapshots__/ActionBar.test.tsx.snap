// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ActionBar > renders 1`] = `
<DocumentFragment>
  <div
    class="flmu0va"
    data-testid="action-bar"
  >
    <button
      class="f1eue72v"
      data-testid="date-picker-button"
    >
      <div
        class="f13oks9e"
        data-testid="icon"
      >
        <svg
          fill="none"
          height="21"
          viewBox="0 0 21 21"
          width="21"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.625 0.5C8.9375 0.5 9.25 0.8125 9.25 1.125V3H15.5V1.125C15.5 0.8125 15.7734 0.5 16.125 0.5C16.4375 0.5 16.75 0.8125 16.75 1.125V3H18C19.3672 3 20.5 4.13281 20.5 5.5V8V14.25C20.5 15.6562 19.3672 16.75 18 16.75H6.75C5.34375 16.75 4.25 15.6562 4.25 14.25V8V5.5C4.25 4.13281 5.34375 3 6.75 3H8V1.125C8 0.8125 8.27344 0.5 8.625 0.5ZM6.75 4.25C6.04688 4.25 5.5 4.83594 5.5 5.5V6.75H19.25V5.5C19.25 4.83594 18.6641 4.25 18 4.25H6.75ZM19.25 8H5.5V14.25C5.5 14.9531 6.04688 15.5 6.75 15.5H18C18.6641 15.5 19.25 14.9531 19.25 14.25V8ZM1.75 8.625V15.5C1.75 17.5703 3.42969 19.25 5.5 19.25H14.875C15.1875 19.25 15.5 19.5625 15.5 19.875C15.5 20.2266 15.1875 20.5 14.875 20.5H5.5C2.72656 20.5 0.5 18.2734 0.5 15.5V8.625C0.5 8.3125 0.773438 8 1.125 8C1.4375 8 1.75 8.3125 1.75 8.625Z"
            fill="currentcolor"
          />
        </svg>
      </div>
      <div
        class="text"
      >
        <div
          class="label"
        >
          <label>
            Date
          </label>
          <svg
            fill="none"
            height="7"
            viewBox="0 0 12 7"
            width="12"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5.60156 5.89844L1.10156 1.39844C0.867188 1.1875 0.867188 0.835938 1.10156 0.601562C1.3125 0.390625 1.66406 0.390625 1.89844 0.601562L6 4.72656L10.1016 0.625C10.3125 0.390625 10.6641 0.390625 10.8984 0.625C11.1094 0.835938 11.1094 1.1875 10.8984 1.39844L6.375 5.89844C6.16406 6.13281 5.8125 6.13281 5.60156 5.89844Z"
              fill="currentcolor"
            />
          </svg>
        </div>
        June 26, 2023
      </div>
    </button>
    <div
      class="fmkmivb"
      data-testid="divider"
    />
    <button
      class="f1eue72v"
      data-testid="num-results-picker-button"
    >
      <div
        class="f13oks9e"
        data-testid="icon"
      >
        <svg
          fill="none"
          height="15"
          viewBox="0 0 20 15"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M2 0C2.66406 0 3.25 0.585938 3.25 1.25C3.25 1.95312 2.66406 2.5 2 2.5C1.29688 2.5 0.75 1.95312 0.75 1.25C0.75 0.585938 1.29688 0 2 0ZM6.375 0.625H18.875C19.1875 0.625 19.5 0.9375 19.5 1.25C19.5 1.60156 19.1875 1.875 18.875 1.875H6.375C6.02344 1.875 5.75 1.60156 5.75 1.25C5.75 0.9375 6.02344 0.625 6.375 0.625ZM6.375 6.875H18.875C19.1875 6.875 19.5 7.1875 19.5 7.5C19.5 7.85156 19.1875 8.125 18.875 8.125H6.375C6.02344 8.125 5.75 7.85156 5.75 7.5C5.75 7.1875 6.02344 6.875 6.375 6.875ZM6.375 13.125H18.875C19.1875 13.125 19.5 13.4375 19.5 13.75C19.5 14.1016 19.1875 14.375 18.875 14.375H6.375C6.02344 14.375 5.75 14.1016 5.75 13.75C5.75 13.4375 6.02344 13.125 6.375 13.125ZM3.25 7.5C3.25 8.20312 2.66406 8.75 2 8.75C1.29688 8.75 0.75 8.20312 0.75 7.5C0.75 6.83594 1.29688 6.25 2 6.25C2.66406 6.25 3.25 6.83594 3.25 7.5ZM2 12.5C2.66406 12.5 3.25 13.0859 3.25 13.75C3.25 14.4531 2.66406 15 2 15C1.29688 15 0.75 14.4531 0.75 13.75C0.75 13.0859 1.29688 12.5 2 12.5Z"
            fill="currentcolor"
          />
        </svg>
      </div>
      <div
        class="text"
      >
        <div
          class="label"
        >
          <label>
            Num Results
          </label>
          <svg
            fill="none"
            height="7"
            viewBox="0 0 12 7"
            width="12"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5.60156 5.89844L1.10156 1.39844C0.867188 1.1875 0.867188 0.835938 1.10156 0.601562C1.3125 0.390625 1.66406 0.390625 1.89844 0.601562L6 4.72656L10.1016 0.625C10.3125 0.390625 10.6641 0.390625 10.8984 0.625C11.1094 0.835938 11.1094 1.1875 10.8984 1.39844L6.375 5.89844C6.16406 6.13281 5.8125 6.13281 5.60156 5.89844Z"
              fill="currentcolor"
            />
          </svg>
        </div>
        100
      </div>
    </button>
    <div
      class="fmkmivb"
      data-testid="divider"
    />
    <button
      class="f1eue72v"
      data-testid="language-picker-button"
    >
      <div
        class="f13oks9e"
        data-testid="icon"
      >
        <svg
          fill="none"
          height="21"
          viewBox="0 0 21 21"
          width="21"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M10.5 19.25C11.125 19.25 12.0625 18.7031 12.8828 17.0234C13.2734 16.2422 13.625 15.3047 13.8594 14.25H7.14062C7.33594 15.3047 7.6875 16.2422 8.07812 17.0234C8.89844 18.7031 9.83594 19.25 10.5 19.25ZM6.90625 13H14.0547C14.1719 12.2188 14.25 11.3984 14.25 10.5C14.25 9.64062 14.1719 8.82031 14.0547 8H6.90625C6.78906 8.82031 6.75 9.64062 6.75 10.5C6.75 11.3984 6.78906 12.2188 6.90625 13ZM7.14062 6.75H13.8594C13.625 5.69531 13.2734 4.79688 12.8828 4.01562C12.0625 2.33594 11.125 1.75 10.5 1.75C9.83594 1.75 8.89844 2.33594 8.07812 4.01562C7.6875 4.79688 7.33594 5.69531 7.14062 6.75ZM15.3438 8C15.4219 8.82031 15.5 9.64062 15.5 10.5C15.5 11.3984 15.4219 12.2188 15.3438 13H18.8594C19.0938 12.2188 19.25 11.3984 19.25 10.5C19.25 9.64062 19.0938 8.82031 18.8594 8H15.3438ZM18.3906 6.75C17.375 4.64062 15.5391 3 13.2734 2.21875C14.0938 3.3125 14.7578 4.91406 15.1094 6.75H18.3906ZM5.85156 6.75C6.24219 4.91406 6.86719 3.3125 7.6875 2.21875C5.42188 3 3.58594 4.64062 2.57031 6.75H5.85156ZM2.10156 8C1.86719 8.82031 1.75 9.64062 1.75 10.5C1.75 11.3984 1.86719 12.2188 2.10156 13H5.65625C5.53906 12.2188 5.5 11.3984 5.5 10.5C5.5 9.64062 5.53906 8.82031 5.65625 8H2.10156ZM13.2734 18.8203C15.5391 18.0391 17.375 16.3984 18.3906 14.25H15.1094C14.7578 16.125 14.0938 17.6875 13.2734 18.8203ZM7.6875 18.8203C6.86719 17.6875 6.24219 16.125 5.85156 14.25H2.57031C3.58594 16.3984 5.42188 18.0391 7.6875 18.8203ZM10.5 20.5C6.90625 20.5 3.625 18.625 1.82812 15.5C0.03125 12.4141 0.03125 8.625 1.82812 5.5C3.625 2.41406 6.90625 0.5 10.5 0.5C14.0547 0.5 17.3359 2.41406 19.1328 5.5C20.9297 8.625 20.9297 12.4141 19.1328 15.5C17.3359 18.625 14.0547 20.5 10.5 20.5Z"
            fill="currentcolor"
          />
        </svg>
      </div>
      <div
        class="text"
      >
        <div
          class="label"
        >
          <label>
            Language
          </label>
          <svg
            fill="none"
            height="7"
            viewBox="0 0 12 7"
            width="12"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5.60156 5.89844L1.10156 1.39844C0.867188 1.1875 0.867188 0.835938 1.10156 0.601562C1.3125 0.390625 1.66406 0.390625 1.89844 0.601562L6 4.72656L10.1016 0.625C10.3125 0.390625 10.6641 0.390625 10.8984 0.625C11.1094 0.835938 11.1094 1.1875 10.8984 1.39844L6.375 5.89844C6.16406 6.13281 5.8125 6.13281 5.60156 5.89844Z"
              fill="currentcolor"
            />
          </svg>
        </div>
      </div>
    </button>
    <button
      class="f1g1amr8"
      data-testid="search-button"
    >
      Search
    </button>
  </div>
</DocumentFragment>
`;
